### PR TITLE
Fills in some of our testing philosophy

### PR DESCRIPTION
**Pull request by swilliams**
Based on some of the conversations we've had. Open to suggestions/edits.

### CODE REVIEW

File: docs/Quality.md [hunk 1]
```diff
@@ -6,16 +6,47 @@ title: "Tallwave Guidelines — Quality"
...
+There is a delicate balance between doing great work and actually shipping something. We can take years to build something with a beautiful and clever architecture and no technical debt, but it would be a misuse of our clients' budgets and time to do so. Likewise, we can sub-contract out to the cheapest off-shore bidders and ship something missing key features that crashes when you look at it funny.
 
-## Tests
+We do believe it is possible to ship Good code in a reasonable amount of time, without gouging on price.
+
+## Testing
+This is our first line of defense. While it is possible to go overboard with testing zealotry, there is great value in _automated_ tests that run against your code. Tests fall into different categories. 
 
 ### Unit Tests
+A unit test examines a single function or method (a single "unit"). For example, the tests for a function called `addIntegers` would make sure that the results of that function are the sum of the arguments passed in. These are the easiest to write, but also provide the least amount of utility. More bugs occur when the units meet, or to use the proper nomenclature, when they integrate with each other.
```
Comment (TinaHeiligers) [inline docs/Quality.md]: ", when they integrate with each other." implies only two units working together. ', when they integrate with one another.' implies many units interacting.

Comment (TinaHeiligers) [inline docs/Quality.md]: ", when they integrate with each other." implies only two units working together. ', when they integrate with one another.' implies many units interacting.

File: docs/Quality.md [hunk 2]
```diff
@@ -6,16 +6,47 @@ title: "Tallwave Guidelines — Quality"
...
 
 ### Integration Tests
+A proper unit test stubs out or mocks dependencies on the function. An integration test runs through 2 or more of those units and checks to see if the result is what was expected.
```
Comment (TinaHeiligers) [inline docs/Quality.md]: '.. if the result is what is expected.' Tests continue to run and the result must continue to be what is expected at the time of running the test. (semantics: past or present tense).

File: docs/Quality.md [hunk 3]
```diff
@@ -6,16 +6,47 @@ title: "Tallwave Guidelines — Quality"
...
 
-### Coverage
+Because the tests depend on more pieces of code they tend to be harder to maintain. Integration tests usually don't go through the UI, though they may interact with the persistance layer. This requires a test environment to be running so that other environments don't get filled with test data.
```
Comment (TinaHeiligers) [inline docs/Quality.md]: 'Harder' or 'more difficult'. I prefer the latter.

Comment (TinaHeiligers) [reply docs/Quality.md]: '...get filled in...' or '...get polluted...'

File: docs/Quality.md [hunk 4]
```diff
@@ -6,16 +6,47 @@ title: "Tallwave Guidelines — Quality"
...
 
+### Full End to End Tests
+To take it a step further, we can use something like [Selenium](http://www.seleniumhq.org) to perform actual UI tasks and then check the whole system to see if its state is what we expected. E2E tests can provide a significant amount of value, but are also the hardest to write and maintain.
```
Comment (TinaHeiligers) [inline docs/Quality.md]: '...what we expect.' Past or present tense. 
'...are the most difficult to write...'

File: docs/Quality.md [hunk 5]
```diff
@@ -6,16 +6,47 @@ title: "Tallwave Guidelines — Quality"
...
+A proper unit test stubs out or mocks dependencies on the function. An integration test runs through 2 or more of those units and checks to see if the result is what was expected.
 
-### Coverage
+Because the tests depend on more pieces of code they tend to be harder to maintain. Integration tests usually don't go through the UI, though they may interact with the persistance layer. This requires a test environment to be running so that other environments don't get filled with test data.
 
+### Full End to End Tests
+To take it a step further, we can use something like [Selenium](http://www.seleniumhq.org) to perform actual UI tasks and then check the whole system to see if its state is what we expected. E2E tests can provide a significant amount of value, but are also the hardest to write and maintain. 
+
+### Coverage
+Utilities like [Istanbul](https://github.com/gotwarlost/istanbul) can tell you the percentage of your codebase that your automated tests hit. This is a useful statistic, but it is just that, a data point, and should not be chased just to hit an arbitrary metric. It is still possible to have bugs in a codebase that has 100% coverage.
```
Comment (TinaHeiligers) [inline docs/Quality.md]: 'Utilities like Istanbul tell you the...'. These utilities not only can but _do_ let you know the % coverage.

File: docs/Quality.md [hunk 6]
```diff
@@ -6,16 +6,47 @@ title: "Tallwave Guidelines — Quality"
...
+
+### Testing Libraries
+* React (including React Native)
+	* [Jest](http://facebook.github.io/jest/)
+* Angular
+	* [Jasmine](https://jasmine.github.io) & [Karma](https://karma-runner.github.io/1.0/index.html) — The preferred combo as it comes with [Angular CLI](https://cli.angular.io)
+	* [Mocha](https://mochajs.org) & [Chai](http://chaijs.com) — is a more generic test runner, but requires more manual work it integrate with Angular
```
Comment (TinaHeiligers) [inline docs/Quality.md]: '...but requires more work _to_ integrate it with Angular.'

File: docs/Quality.md [hunk 7]
```diff
@@ -6,16 +6,47 @@ title: "Tallwave Guidelines — Quality"
...
+* Node
+	* Mocha & Chai
+* iOS/Swift
+	* [xctest](https://developer.apple.com/documentation/xctest) — There are other libraries to provide a more behavioral driven testing approach, but those are at the mercy of Xcode's tempermentalness.
```
Comment (TinaHeiligers) [inline docs/Quality.md]: 'There are other libraries that provide a more behavioral...'

File: docs/Quality.md (hunk 1 above)
Comment (swilliams) [reply docs/Quality.md]: I get that one wrong a lot.

**Comment by swilliams** [comment]
Thank you @vinnieking06 and @TinaHeiligers for the feedback. I've implemented your recommendations.

**Comment by swilliams** [comment]
[This isn't done, please add and submit a PR to add something else]

**Pull request merged**